code/datums/components/food/ice_cream_holder.dm

```diff
@@ -0,0 +1,280 @@
+	src.change_name = change_name
+	src.filled_name = filled_name
+	src.change_desc = change_desc
+	src.x_offset = x_offset
+	src.y_offset = y_offset
+	src.sweetener = sweetener
+
+	var/update_flags = NONE
+	RegisterSignal(owner, COMSIG_ITEM_ATTACK_OBJ, .proc/on_item_attack_obj)
+	RegisterSignal(owner, COMSIG_ICE_CREAM_ADD_SERVING, .proc/add_serving)
+	RegisterSignal(owner, COMSIG_ATOM_UPDATE_OVERLAYS, .proc/on_update_overlays)
+	if(change_name)
+		RegisterSignal(owner, COMSIG_ATOM_UPDATE_NAME, .proc/on_update_name)
+		update_flags |= UPDATE_NAME
+	if(!change_desc)
+		RegisterSignal(owner, COMSIG_PARENT_EXAMINE_MORE, .proc/on_examine_more)
+	else
+		RegisterSignal(owner, COMSIG_ATOM_UPDATE_DESC, .proc/on_update_desc)
+		update_flags |= UPDATE_DESC
+	if(update_flags)
+		owner.update_appearance(update_flags)
+
+/datum/component/ice_cream_holder/proc/on_update_name(atom/source, updates)
+	var/obj/obj = source
+	if(istype(obj) && obj.renamedByPlayer) //Renamed by the player.
+		return
+	var/scoops_len = length(scoops)
+	if(!scoops_len)
+		source.name = initial(source.name)
+	else
+		var/name_to_use = filled_name || initial(source.name)
+		var/list/unique_list = uniqueList(scoops)
+		if(scoops_len > 1 && length(unique_list) == 1) // multiple flavours, and all of the same type
+			source.name = "[make_tuple(scoops_len)] [scoops[1]] [name_to_use]" // "double vanilla" sounds cooler than just "vanilla"
+		else
+			source.name = "[english_list(unique_list)] [name_to_use]"
+
+/datum/component/ice_cream_holder/proc/on_update_desc(atom/source, updates)
+	var/obj/obj = source
+	if(istype(obj) && obj.renamedByPlayer) //Renamed by the player.
+		return
+	var/scoops_len = length(scoops)
+	if(!scoops_len)
+		source.desc = initial(source.desc)
+	else if(scoops_len == 1 || length(uniqueList(scoops)) == 1) /// Only one flavour.
+		var/key = scoops[1]
+		var/datum/ice_cream_flavour/flavour = GLOB.ice_cream_flavours[LAZYACCESS(special_scoops, key) || key]
+		if(!flavour?.desc) //I scream.
+			source.desc = initial(source.desc)
+		else
+			source.desc = replacetext(replacetext(flavour.desc, "$CONE_NAME", initial(source.name)), "$CUSTOM_NAME", key)
+	else /// Many flavours.
+		source.desc = "A delicious [initial(source.name)] filled with scoops of [english_list(scoops)] icecream. That's as many as [scoops_len] scoops!"
+
+/datum/component/ice_cream_holder/proc/on_examine_more(atom/source, mob/mob, list/examine_list)
+	var/scoops_len = length(scoops)
+	if(scoops_len == 1 || length(uniqueList(scoops)) == 1) /// Only one flavour.
+		var/key = scoops[1]
+		var/datum/ice_cream_flavour/flavour = GLOB.ice_cream_flavours[LAZYACCESS(special_scoops, key) || key]
+		if(flavour?.desc) //I scream.
+			examine_list += "[source.p_theyre(TRUE)] filled with scoops of [flavour ? flavour.name : "broken, unhappy"] icecream."
+		else
+			examine_list += replacetext(replacetext("[source.p_theyre(TRUE)] [flavour.desc]", "$CONE_NAME", initial(source.name)), "$CUSTOM_NAME", key)
```

Comment:
Unrelated to the pr but we could use some generic madlibs helpers.

Reply:
I suppose, though I'd like an example later to get a better idea of this suggestion.

Reply:
Oh don't worry, this is just idle thought. I was thinking some simple `adlib(text,replacement_dictionary/datum)`